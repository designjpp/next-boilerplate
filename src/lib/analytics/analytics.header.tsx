import Script from 'next/script';
import React from 'react';

import type { FC } from 'react';

type AnalyticsHeadScriptProps = {
	url?: string;
};

export const AnalyticsHeadScript: FC<AnalyticsHeadScriptProps> = ({ url }) => {
	if (!url) {
		return null;
	}

	return (
		<Script
			defer
			async
			src={url}
			id='analytics-script'
			data-testid='analytics-script'
			strategy='afterInteractive'
		/>
	);
};
