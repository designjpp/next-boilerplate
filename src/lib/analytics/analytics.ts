import { useEffect, useRef } from 'react';

import { analyticEvent } from './_analytics';

import type { DependencyList, RefObject } from 'react';

/**
 ** useAnalytics

 Please read this document if you are new to the topic:
 https://github.com/movahedan/next-boilerplate/blob/main/docs/directories.md

 * A hook util to send analytic events, that gives the access of analytic command
 * and works in 2 way for manual (dom interaction) and functional tracking
 *
 * @parameters
 * @deps Array must be given (needed to be in additionalHooks in eslint config)
 * @callback Function Control whether it is disabled, manual or functional
 ** if callback returns @false => the request will be disabled
 ** if callback returns an @AnalyticEventInfo => the request will be send functionally
 ** if callback returns an @AnalyticTracker => the ref will be yours to do the request manually
 *
 * @returns
 * @A_ref object that you can bind to an Element when using manual tracking
 * If you not use it and want a manual track, it will throw an Error
 *
 * @example of functional tracking
 ** useAnalytics(() => getSamplePageAnalyticsProps(), []);
 *
 * return (
 *   <div>
 *   </div>
 * );
 *
 * @example of manual tracking
 ** const ref = useAnalytics(() => trackHeaderInteractions, []);
 *
 * return (
 *   <div ref={ref} id="id">
 *   </div>
 * );
 *
 * @example of disabled
 ** useAnalytics(() => false, []);
 *
 * return (
 *   <div>
 *   </div>
 * );
 *
 */

export type AnalyticEventInfo<
	AnalyticEventType = string,
	AnalyticEventData = unknown
> = {
	eventType: AnalyticEventType;
	data: AnalyticEventData;
};
export type AnalyticTracker<ContainerElement = HTMLDivElement> = (
	containerRef: RefObject<ContainerElement>,
	analyticCommand: typeof analyticEvent
) => void;

export const useAnalytics = <
	AnalyticEventType = string,
	AnalyticEventData = unknown,
	ContainerElement = HTMLDivElement
>(
	callback: () =>
		| false
		| AnalyticEventInfo<AnalyticEventType, AnalyticEventData>
		| AnalyticTracker<ContainerElement>,
	deps: DependencyList
): RefObject<ContainerElement> => {
	const containerRef = useRef<ContainerElement>(null);

	useEffect(() => {
		const result = callback();

		if (result) {
			if (typeof result === 'function') {
				if (containerRef.current === null) {
					throw Error(
						'[useAnalytics]: containerRef is not attached to any element'
					);
				}

				result(containerRef, analyticEvent);
			} else {
				const { eventType, data } = result;
				analyticEvent<AnalyticEventType, AnalyticEventData>(eventType, data);
			}
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, deps);

	return containerRef;
};
