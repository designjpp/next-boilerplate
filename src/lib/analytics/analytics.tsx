import { useAnalyticsPageEvent, useAuthenticatedEvent } from './events';

import type { FC } from 'react';

export const AnalyticsProvider: FC = () => {
	useAnalyticsPageEvent();
	useAuthenticatedEvent();

	return null;
};
