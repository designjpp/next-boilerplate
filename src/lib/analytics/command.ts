export const analyticEvent = <
	AnalyticEventType = string,
	AnalyticEventData = unknown
>(
	name: AnalyticEventType,
	data: AnalyticEventData
): void => {
	if (name && global.window) {
		global.window.dataLayer = global.window.dataLayer || [];
		global.window.dataLayer.push({
			event: name,
			data: data,
		});
	}
};
