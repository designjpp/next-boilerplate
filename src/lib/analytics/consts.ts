export const analyticsId = process.env.NEXT_PUBLIC_ANALYTICS_ID;
export const analyticsTrackingId =
  process.env.NEXT_PUBLIC_ANALYTICS_TRACKING_ID;
