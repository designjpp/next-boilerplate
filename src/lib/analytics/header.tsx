import React from 'react';

import type { FC } from 'react';

type AnalyticsHeadScriptProps = {
	url?: string;
};

export const AnalyticsHeadScript: FC<AnalyticsHeadScriptProps> = ({ url }) => {
	if (!url) {
		return null;
	}

	return <script data-testid='analytics-script' src={url} async defer></script>;
};
