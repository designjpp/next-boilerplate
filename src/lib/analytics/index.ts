export * from './analytics';
export * from './analytics.header';
