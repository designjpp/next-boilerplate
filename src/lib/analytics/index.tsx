import { useAnalyticsPageEvent, useAuthenticatedEvent } from './providers';

import type { FC } from 'react';

export const AnalyticsProvider: FC = () => {
	useAnalyticsPageEvent();
	useAuthenticatedEvent();

	return null;
};

export * from './script';
export * from './events';
