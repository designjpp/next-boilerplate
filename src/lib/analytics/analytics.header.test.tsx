import { render } from '@testing-library/react';
import { create } from 'react-test-renderer';

import { AnalyticsHeadScript } from 'lib/analytics';

describe('<AnalyticsHeadScript>', () => {
	it('should renders correct script', async () => {
		const url = 'https://domain.com';
		const { queryByTestId } = render(<AnalyticsHeadScript url={url} />);

		const scriptElement = queryByTestId('analytics-script');

		expect(scriptElement).toHaveAttribute('async');
		expect(scriptElement).toHaveAttribute('defer');
		expect(scriptElement).toHaveAttribute('src', url);
	});

	it('should returns null when no url is passed to it', async () => {
		const component = create(<AnalyticsHeadScript />).toJSON();

		expect(component).toMatchSnapshot();
	});
});
