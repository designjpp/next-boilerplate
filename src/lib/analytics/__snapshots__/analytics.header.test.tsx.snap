// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnalyticsHeadScript> should returns null when no url is passed to it 1`] = `null`;
